ldap/servers/plugins/replication/cl5_api.c

```diff
@@ -2291,142 +2545,81 @@ _cl5DoPurging(cleanruv_purge_data *purge_data)
     return;
 }
 
+static inline int
+_cl5CIEventCheckTxnEnd(DBLCI_EVENT_COUNT *ev)
+{
+    return (ev->nbmax && ev->nb >= ev->nbmax);
+}
+
 /*
- * If the rid is not set it is the very first iteration of the changelog.
- * If the rid is set, we are doing another pass, and we have a key as our
- * starting point.
+ * _cl5Iterate callbacks helper
+ * Changelog cursor iterator callback common code initializer.
+ *  Shoukd start all the _cl5CursorIterate callbacks.
  */
 static int
-_cl5PurgeGetFirstEntry(cldb_Handle *cldb, CL5Entry *entry, void **iterator, dbi_txn_t *txnid, int rid, dbi_val_t *key)
+_cl5CICbInit(dbi_val_t *key, dbi_val_t *data, DBLCI_CTX *dblcictx)
 {
-    dbi_cursor_t cursor = {0};
-    dbi_val_t data = {0};
-    CL5Iterator *it;
-    int rc;
+    /* Verify that we have a valid csn */
+    if (key->size != CSN_STRSIZE) {
+        return DBI_RC_SUCCESS;
+    }
+    /* Update last csn */
+    csn_init_by_string(&dblcictx->csn, data->data);
+    if (_cl5CIEventCheckTxnEnd(&dblcictx->seen) ||
+        _cl5CIEventCheckTxnEnd(&dblcictx->changed)) {
+        /* Stop this txn */
+        return DBI_RC_NOTFOUND;
```

Comment:
As the txn overpass a limit (like CL5_PURGE_MAX_LOOKUP_PER_TRANSACTION or CL5_PURGE_MAX_PER_TRANSACTION) shouldn't we return DBI_RC_RETRY ?

Reply:
No: because in retry case, the txn is aborted and iteration is restarted from the starting point.
but with DBI_RC_NOTFOUND the dblayer_iterate function  stops with a SUCCESS rc 
then the txn is commited and the cl5_iterate function restarts a new iteration on last seen record

Reply:
Excellent. Thanks for these explanations. Do you mind to add this comment (DBI_RC_NOTFOUND->dblayer_iterate succeeds -> txn comming..). It could help in the future.

Reply:
OK. I added a comment